components/user-mgt/org.wso2.carbon.identity.user.pre.update.profile.action/src/main/java/org/wso2/carbon/identity/user/pre/update/profile/action/internal/execution/PreUpdateProfileRequestBuilder.java

```diff
@@ -0,0 +1,327 @@
+                    UserCoreConstants.DEFAULT_PROFILE);
+
+            String multiAttributeSeparator = FrameworkUtils.getMultiAttributeSeparator();
+            setClaimsInUserBuilder(userBuilder, claimValues, userActionRequestDTO.getClaims(), multiAttributeSeparator);
+            setGroupsInUserBuilder(userBuilder, claimValues, multiAttributeSeparator);
+        } catch (org.wso2.carbon.user.core.UserStoreException e) {
+            throw new ActionExecutionRequestBuilderException("Failed to retrieve user claims from user store.", e);
+        }
+        return userBuilder.build();
+    }
+
+    private void setClaimsInUserBuilder(User.Builder userBuilder, Map<String, String> claimValues,
+                                        Map<String, Object> updatingUserClaimsInRequest,
+                                        String multiAttributeSeparator) throws ActionExecutionRequestBuilderException {
+
+        List<UpdatingUserClaim> userClaimValuesToSetInEvent = new ArrayList<>();
+        for (Map.Entry<String, String> claimEntry : claimValues.entrySet()) {
+            String claimKey = claimEntry.getKey();
+            String claimValue = claimEntry.getValue();
+
+            if (isRoleOrGroupClaim(claimKey) || StringUtils.isBlank(claimValue)) {
+                continue;
+            }
+
+            if (isMultiValuedClaim(claimKey)) {
+                userClaimValuesToSetInEvent.add(
+                        constructMultiValuedClaim(updatingUserClaimsInRequest, claimKey, claimValue,
+                                multiAttributeSeparator));
+            } else {
+                userClaimValuesToSetInEvent.add(constructSingleValuedClaim(updatingUserClaimsInRequest, claimKey,
+                        claimValue));
+            }
+        }
+
+        userBuilder.claims(userClaimValuesToSetInEvent);
+    }
+
+    private void setGroupsInUserBuilder(User.Builder userBuilder, Map<String, String> claimValues,
+                                        String multiAttributeSeparator) {
+
+        if (claimValues.get(GROUP_CLAIM_URI) != null) {
+            userBuilder.groups(Arrays.asList(
+                    claimValues.get(GROUP_CLAIM_URI).split(Pattern.quote(multiAttributeSeparator))));
+        }
+    }
+
+    private boolean isRoleOrGroupClaim(String claimKey) {
+
+        return ROLE_CLAIM_URI.equals(claimKey) || GROUP_CLAIM_URI.equals(claimKey);
+    }
+
+    private UpdatingUserClaim constructMultiValuedClaim(Map<String, Object> updatingUserClaimsInRequest,
+                                                        String claimKey, String claimValue,
+                                                        String multiAttributeSeparator)
+            throws ActionExecutionRequestBuilderException {
+
+        if (updatingUserClaimsInRequest.containsKey(claimKey)) {
+            Object updatingClaimValue = updatingUserClaimsInRequest.get(claimKey);
+            if (!(updatingClaimValue instanceof String[])) {
```

Comment:
Is it possible for a multi-valued claim to have a single list item as a string in updatingUserClaimsInRequest?

Reply:
No it's not possible. If it's a multi valued attribute the DTO always includes a String[]